### DISCUSSION TITLE
Rasterization

### DISCUSSION THREAD

**Discussion by RobinSchmidt** [RobinSchmidt/RS-MET#332] [Show and tell]
i'm currently looking into rasterization algorithms for drawing primitives such as triangles and polygons. given the ubiquity of the demand for such algorithms in computer graphics, you would expect to find detailed information about that easily on the web. but not so. even my two computer graphics textbooks only scratch at the surface in that area. they do discuss some basic techniques for non-anti-aliased rasterization - but anti-aliased algos are only superficially outlined (and fully discussed only for lines with single-pixel thickness). this lack of info is also noted here:

https://www.scratchapixel.com/lessons/3d-basic-rendering/rasterization-practical-implementation/overview-rasterization-algorithm

"The rasterization rendering technique is surely the most commonly used technique to render images of 3D scenes, and yet, that is probably the least understood and the least properly documented technique of all"

so i'm trying to come up with an algorithm for rasterizing triangles with anti-aliasing. here are some (triangulated) polygons placed on a pixel grid:

![image](https://user-images.githubusercontent.com/1833099/42411029-b665bca8-81f4-11e8-9e50-e146382bf81b.png)

here's a rendering via a non-anti-aliased raterization function modeled after that:
https://www.youtube.com/watch?v=9A5TVh6kPLA

![image](https://user-images.githubusercontent.com/1833099/42411036-d6259388-81f4-11e8-8f14-68cf1f5a8c4a.png)

which looks pretty ugly, even though modeled after some of the most detailed information that could be easily found on the web. now here is my current prototype for anti-aliased rasterization:

![image](https://user-images.githubusercontent.com/1833099/42411037-dade879a-81f4-11e8-85a6-90709227e653.png)

**Comment by RobinSchmidt** [comment]
...and even though, at scratchapixel, they say:

"With this lesson, we believe to be the first ressource that provides a clear and complete picture of the algorithm as well as a complete and full practical implementation of the technique"

they also only discuss rasterization *without* anti-aliasing - here:

https://www.scratchapixel.com/lessons/3d-basic-rendering/rasterization-practical-implementation/rasterization-stage

...really weird. looks like i had to re-invent the wheel once again. libraries, such as AGG (http://antigrain.com/) also do not really help very much - it just uses "supersampling" (computer graphics community lingo for what we audio guys call oversampling) - which is basically throwing the towel, imho. i have implemented what is called unweighted area sampling, described here:

http://www.cs.cmu.edu/afs/cs/academic/class/15462-s16/www/lec_slides/2.pdf

my current prototype implementation is ultra-expensive. it involves computing the clipped polygons resulting from clipping the triangle to each pixel's square (via sutherland-hodgman algorithm) and then computing the area of that polygon. this gives the exact pixel-coverage values. good enough for a proof of concept but not practical, when performance is required. 

next step would be to make it more practical by computing coverage values only where actually needed (in areas of zero or full coverage, i don't need to compute it) and getting rid of the full blown sutherland-hodgman polygon clipping algo by taking advantage of the simple geometries of the two polygons involved (a triangle and a square - the current algorithm can clip any (even nonconvex) polygon against a convex clipping polygon - and is quite expensive). maybe i'll try to get that working and then move on to the meta-parameter mapping

**Comment by elanhickler** [comment]
what are you planning on doing with this?

**Comment by RobinSchmidt** [comment]
good question. i have nothing particular in mind just yet. i just think, it may be a useful thing to have in my toolbag. rasterizing triangles is the most low-level operation in any (rasterization-based (*)) graphics rendering system, because any more complex geometry will at some level be broken down into triangles ("triangulation"). and as said - information about how to do this fundamental step *right* (and without cheating like in supersampling) is hard to find and i want to save the technique from getting lost. it's so low-level, that it tends to be provided by the graphics hardware, so nobody - besides graphics hardware manufacturers - seems to care about that anymore. and they keep their algorithms under the wraps and everybody else takes them for granted. but obviously, to code a renderer from scratch, you will need code for that step. maybe at some point, i may do some higher level operations on top of that to do an actual renderer (but *much* more stuff will be required for that).

(*) there's a complementary approach to rasterization that seems to be in more widespread use today in any serious renderer - namely raytracing. i guess, it's more straightforward to incorporate physically correct visual effects with this approach which is why it has become state-of-art (but that's just a guess). the main difference is that in  rasterization, you have an outer loop over the triangles and for each triangle an inner loop over the pixels of the triangle and in raytracing your outer loop runs over all pixels and for each pixel, the algorithm computes intersections of the ray through the pixel with all triangles. so inner and outer loops are basically swapped. rasterization has the advantage compared to raytracing that it first breaks down a 3D geometry to a 2D geometry and the rasterization is actually applied to the 2D geometry. raytracing works directly in 3D. it avoids any intermediate 2D geometry stage. but if you want to just render 2D geometry, raytracing seems a totally over-the-top and awkward approach (it certainly can be done - you'd just have to artificially embed a 2D plane into 3D space and just place all your 3D triangles flatly onto that plane - but that would be a really awkward thing to do)

however, i have still a few cases, where my optimized algorithm fails and i need a break from that now, so i will now look into meta-mapping. that seems to be - as far as i can see - a general pattern for geometry code: you have to distinguish a shitload of different cases and treat each a little bit different, requiring messy code with complex nested conditionals - very different to what i'm used to from audio code. audio processing code tends to be fairly linear - mostly serial, sometimes parallel. but graphics code tends to branch out into a complicated tree of different code-paths. my graphics code looks very different from my audio code...lots of "if"s and "case"s

**Comment by RobinSchmidt** [comment]
i just discovered that i can assign labels to the different threads - i actually didn't mean to bump this up. ...might be a useful feature because the board can also filter threads based on this
